### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by m26jiang**

### CODE REVIEW

File: app/persona/api/authentication/AuthenticationServiceImpl.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.UUID
+
+import com.google.inject.Inject
+import com.mohiva.play.silhouette.api.LoginInfo
+import com.mohiva.play.silhouette.impl.providers.CommonSocialProfile
+import persona.model.authentication.User
+import persona.model.dao.authentication.UserDAO
+
+import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (taylorstark) [inline app/persona/api/authentication/AuthenticationServiceImpl.scala]: As per previous comment, consider import play.api.libs.concurrent.Execution.Implicits._.  Read the documentation here: https://www.playframework.com/documentation/2.4.x/ThreadPools.  Decide if the work you're doing is blocking or non-blocking.  If it's non-blocking, switch to the play thread pool.  If it's blocking, switch to the play thread pool and open a ticket reminding us to do more investigation in the future

File: app/persona/api/authentication/AuthenticationServiceImpl.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import scala.concurrent.Future
+
+class AuthenticationServiceImpl @Inject()(userDAO: UserDAO) extends AuthenticationService {
+
+  override def save(user: User): Future[User] = userDAO.save(user)
+
+  override def save(profile: CommonSocialProfile): Future[User] = {
+    userDAO.find(profile.loginInfo).flatMap {
+      case Some(user) =>
```
Comment (taylorstark) [inline app/persona/api/authentication/AuthenticationServiceImpl.scala]: What's the purpose of saving someone that's already in our database?  Pretty sure I'm just not understanding what CommonSocialProfile does, not that the code is wrong

File: app/persona/controllers/Authentication.scala [hunk 3]
```diff
@@ -2,12 +2,49 @@ package persona.controllers
...
+    credentialsProvider.authenticate(request.body).flatMap { loginInfo =>
+      authenticationService.retrieve(loginInfo).flatMap {
+        case Some(user) => env.authenticatorService.create(user.loginInfo).flatMap { authenticator =>
+          env.eventBus.publish(LoginEvent(user, request, Messages(request2lang, messagesApi)))
+          env.authenticatorService.init(authenticator).flatMap { cookie =>
+            env.authenticatorService.embed(cookie, Ok("").withCookies(cookie))
+          }
+        }
+        case None =>
+          Future.failed(new IdentityNotFoundException("User not found"))
```
Comment (taylorstark) [inline app/persona/controllers/Authentication.scala]: Probably shouldn't throw an exception

File: app/persona/api/authentication/AuthenticationServiceImpl.scala (hunk 2 above)
Comment (m26jiang) [inline app/persona/api/authentication/AuthenticationServiceImpl.scala]: User could have multiple linked profiles.
i.e. I could login through Facebook, Google, or through our own service.

File: app/persona/controllers/Authentication.scala (hunk 3 above)
Comment (m26jiang) [inline app/persona/controllers/Authentication.scala]: Silhouette handles the exception. Maybe something to investigate into

**Comment by m26jiang** [comment]
Oops, forgot to update your tests... Dem dependencies

**Comment by m26jiang** [comment]
Rebased with local master..... I'm stupid. Third times the charm

**Comment by rpcrezel** [comment]
lgtm

**Comment by taylorstark** [comment]
LGTM

**Pull request merged**